db/crud.go

```diff
@@ -1989,6 +1980,51 @@ func (db *Database) updateAndReturnDoc(docid string, allowImport bool, expiry ui
 	return doc, newRevID, nil
 }
 
+func getAttachmentIDsForLeafRevisions(db *Database, doc *Document, newRevID string) (map[string]struct{}, error) {
+	leafAttachments := make(map[string]struct{})
+
+	currentAttachments, err := retrieveV2AttachmentKeys(doc.ID, doc.Attachments)
+	if err != nil {
+		return nil, err
+	}
+
+	for attachmentID, _ := range currentAttachments {
+		leafAttachments[attachmentID] = struct{}{}
+	}
+
+	documentLeafRevisions := doc.History.GetLeaves()
+	if len(documentLeafRevisions) > 1 {
+		for _, leafRevision := range documentLeafRevisions {
+			if leafRevision == newRevID {
+				continue
+			}
+
+			revInfo, err := doc.History.getInfo(leafRevision)
```

Comment:
You could use GetLeavesFiltered to only return leaves where `revInfo.HasAttachments && leafRevision != newRevID` - I think that would simplify this code a bit.

Alternately you could use revtree.forEachLeaf to combine the two operations you're doing here (GetLeaves() and getInfo()), and avoid the additional lookup in getInfo.

Reply:
Yeah thats a good plan, implementing #1 seemed to have cleaned it up well